### PR TITLE
2 modules for Packet host: packet_device and packet_sshkey

### PR DESCRIPTION
**Pull request by t0mk**
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
- cloud/packet/packet_device
- cloud/packet/packet_sshkey

##### ANSIBLE VERSION

```
ansible 2.3.0 (packet-modules 8babd1fdda) last updated 2016/12/08 22:27:03 (GMT +300)
  config file = 
  configured module search path = Default w/o overrides
```

##### SUMMARY
This PR adds 2 modules for the Packet host (https://www.packet.net). Packet offers bare metal infrastructure.

- packet_device: for managing devices (servers)
- packet_sshkey: for adding and removing public SSH keys

### CODE REVIEW

File: lib/ansible/modules/cloud/packet/packet_sshkey.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+        'changed': changed,
+        'sshkeys': [serialize_sshkey(k) for k in matching_sshkeys]
+    }
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            state = dict(choices=['present', 'absent'], default='present'),
+            auth_token=dict(default=os.environ.get(PACKET_API_TOKEN_ENV_VAR)),
```
Comment (gundalow) [inline lib/ansible/modules/cloud/packet/packet_sshkey.py]: For secret things please add
`, no_log=True,`

**Comment by t0mk** [comment]
@crunchywelch @baldwinSPC
This the active PR with the modules for Packet host. I made it a single commit.

**Comment by baldwinSPC** [comment]
+1 from this end.

**Comment by t0mk** [comment]
@baldwinSPC thanks!

I tried to fix some things that people pinpointed in your previous PR https://github.com/ansible/ansible-modules-extras/pull/1673:
- idempotent behavious for "present" devices (e.g. new device is not created if one with given specs already exists)
- it's possible to wait for more spawned devices at once (one use-case of this is in examples)

File: lib/ansible/modules/cloud/packet/packet_device.py [hunk 2]
```diff
@@ -0,0 +1,568 @@
...
+
+def is_valid_uuid(myuuid):
+    try:
+        val = uuid.UUID(myuuid, version=4)
+    except ValueError:
+        return False
+    return str(val) == myuuid
+
+
+def listify_string_name_or_id(s):
```
Comment (mmlb) [inline lib/ansible/modules/cloud/packet/packet_device.py:323]: both callers of `listify_string_name_or_id` go on to call `.strip()` on list elements, so maybe you should get rid of that from here?

File: lib/ansible/modules/cloud/packet/packet_device.py [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+    if ',' is s:
```
Comment (mmlb) [inline lib/ansible/modules/cloud/packet/packet_device.py]: typo?: `is` -> `in`

File: lib/ansible/modules/cloud/packet/packet_device.py [hunk 4]
```diff
@@ -0,0 +1,568 @@
...
+    device_data['locked'] = device.locked
+    device_data['ip_addresses'] = [
+        {
+            'address': addr_data['address'],
+            'address_family': addr_data['address_family'],
+            'public': addr_data['public'],
+        }
+        for addr_data in device.ip_addresses
+    ]
+    # Also include each IPs as a key for easier lookup in roles.
```
Comment (mmlb) [inline lib/ansible/modules/cloud/packet/packet_device.py:289]: typo: IPs -> IP

File: lib/ansible/modules/cloud/packet/packet_device.py [hunk 5]
```diff
@@ -0,0 +1,568 @@
...
+  tasks:
+  - packet_device:
+      project_id: 89b497ee-5afc-420a-8fb5-56984898f4df
+      hostnames: server-%02d
+      count: 3
+      operating_system: ubuntu_16_04
+      plan: baremetal_0
+      facility: sjc1
+
+- name: Create 3 coreos devices with userdata, wait till they get IPs and then wait for SSH
```
Comment (mmlb) [inline lib/ansible/modules/cloud/packet/packet_device.py]: grammar: `till` -> `until`

File: lib/ansible/modules/cloud/packet/packet_device.py [hunk 6]
```diff
@@ -0,0 +1,568 @@
...
+
+  state:
+    description:
+      - Desired state of the device.
+    choices: [present, absent, active, inactive, rebooted]
+    default: 'present'
+
+  user_data:
+    description:
+      - Userdata blob made available to the machine
```
Comment (mmlb) [inline lib/ansible/modules/cloud/packet/packet_device.py:93]: missing a `.` at the end

File: lib/ansible/modules/cloud/packet/packet_device.py [hunk 7]
```diff
@@ -0,0 +1,568 @@
...
+    description:
+      - Facility slug for device creation. As of 2016, it should be one of [ewr1, sjc1. ams1].
+
+  features:
+    description:
+      - Dict with "features" for device creation. See Packet API docs for details.
+
+  hostnames:
+    description:
+      - Alias of I(name)
```
Comment (mmlb) [inline lib/ansible/modules/cloud/packet/packet_device.py]: missing `.` at end of sentence

File: lib/ansible/modules/cloud/packet/packet_device.py [hunk 8]
```diff
@@ -0,0 +1,568 @@
...
+    description:
+      - From which number to start the count. 
+
+  device_ids:
+    description:
+      - List of device IDs on which to operate.
+
+  facility:
+    description:
+      - Facility slug for device creation. As of 2016, it should be one of [ewr1, sjc1. ams1].
```
Comment (mmlb) [inline lib/ansible/modules/cloud/packet/packet_device.py]: nrt1 is currently available
also typo after sjc1: `.` -> `,`

File: lib/ansible/modules/cloud/packet/packet_device.py [hunk 9]
```diff
@@ -0,0 +1,568 @@
...
+author: Tomas Karasek <tom.to.the.k@gmail.com>, Matt Baldwin <baldwin@stackpointcloud.com>, Thibaud Morel l'Horset <teebes@gmail.com>
+
+options:
+  auth_token:
+    description:
+      - Packet api token. You can also supply it in env var PACKET_API_TOKEN.
+
+  count:
+    description:
+      - The number of devices to create. Count number can he included in hostname via the %d string formatter.
```
Comment (mmlb) [inline lib/ansible/modules/cloud/packet/packet_device.py]: typo: can he -> can be

File: lib/ansible/modules/cloud/packet/packet_device.py [hunk 10]
```diff
@@ -0,0 +1,568 @@
...
+
+    created_ids = [d.id for d in created_devices]
+    wait_timeout = module.params.get('wait_timeout')
+    wait_timeout = time.time() + wait_timeout
+    while wait_timeout > time.time():
+        refreshed = refresh_created_devices(created_ids, module,
+                                            packet_conn)
+        if all_have_public_ip(refreshed):
+            return refreshed
+            break
```
Comment (mmlb) [inline lib/ansible/modules/cloud/packet/packet_device.py]: unnecessary break

File: lib/ansible/modules/cloud/packet/packet_sshkey.py [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+     - UUID of the key which you want to remove.
+  fingerprint:
+    description:
+     - Fingerprint of the key which you want to remove.
+  key:
+    description:
+     - Public Key string ({type} {base64 encoded key} {description})
+  key_file:
+    description:
+     - File with the public key
```
Comment (mmlb) [inline lib/ansible/modules/cloud/packet/packet_sshkey.py]: missing `.` at end

File: lib/ansible/modules/cloud/packet/packet_sshkey.py [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+     - Label for the key. If you keep it empty, it will be read from key string.
+  id:
+    description:
+     - UUID of the key which you want to remove.
+  fingerprint:
+    description:
+     - Fingerprint of the key which you want to remove.
+  key:
+    description:
+     - Public Key string ({type} {base64 encoded key} {description})
```
Comment (mmlb) [inline lib/ansible/modules/cloud/packet/packet_sshkey.py]: missing `.` at end

**Comment by t0mk** [comment]
Thanks for the comments! I will fix them after Chistmas.

**Comment by siwyd** [comment]
Ran a simple playbook with these modules and it worked nicely for me.

File: lib/ansible/modules/cloud/packet/packet_device.py [hunk 13]
```diff
@@ -0,0 +1,567 @@
...
+    - API is documented at U(https://www.packet.net/help/api/#page:devices,header:devices-devices-post).
+
+version_added: 2.3
+
+author: Tomas Karasek <tom.to.the.k@gmail.com>, Matt Baldwin <baldwin@stackpointcloud.com>, Thibaud Morel l'Horset <teebes@gmail.com>
+
+options:
+  auth_token:
+    description:
+      - Packet api token. You can also supply it in env var PACKET_API_TOKEN.
```
Comment (gundalow) [inline lib/ansible/modules/cloud/packet/packet_device.py]: Environment variables should be formatted in a constant width font:
`C(PACKET_API_TOKEN)`

File: lib/ansible/modules/cloud/packet/packet_device.py (hunk 7 above)
Comment (gundalow) [reply lib/ansible/modules/cloud/packet/packet_device.py]: Please delete this line. 
Any aliases get documented via the `aliases:` option.

File: lib/ansible/modules/cloud/packet/packet_device.py [hunk 14]
```diff
@@ -0,0 +1,567 @@
...
+      - Alias of I(name)
+      - A hostname of a device, or a list of hostnames. 
+      - If given string or one-item list, you can use the C("%d") Python string format to expand numbers from count.
+      - If only one hostname, it might be expanded to list if count>1. 
+    aliases: [name]
+
+  lock:
+    description:
+      - Whether to lock a created device.
+    type: bool
```
Comment (gundalow) [inline lib/ansible/modules/cloud/packet/packet_device.py]: There isn't a `type:` option in the documentation block, please remove.

File: lib/ansible/modules/cloud/packet/packet_sshkey.py [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+    if state in ['present','absent']:
+        try:
+            module.exit_json(**act_on_sshkeys(state, module, packet_conn))
+        except Exception as e:
+            module.fail_json(msg='failed to set sshkey state: %s' % str(e))
+    else:
+        module.fail_json(msg='%s is not a valid state for this module' % state)
+
+
+from ansible.module_utils.basic import * # noqa: F403
```
Comment (gundalow) [inline lib/ansible/modules/cloud/packet/packet_sshkey.py]: Please only import what you need.

**Review by gundalow — CHANGES_REQUESTED** [review]
needs_revision 

Please also update the CHANGELOG.md in the root to list these new modules.

**Comment by t0mk** [comment]
@gundalow: thanks for your comments, I will fix them on Tuesday.

File: lib/ansible/modules/cloud/packet/packet_device.py [hunk 16]
```diff
@@ -35,7 +35,7 @@
 options:
   auth_token:
     description:
-      - Packet api token. You can also supply it in env var PACKET_API_TOKEN.
+      - Packet api token. You can also supply it in env var c(PACKET_API_TOKEN).
```
Comment (gundalow) [inline lib/ansible/modules/cloud/packet/packet_device.py]: Capital `C`, apologies if I told you the wrong thing before.

File: lib/ansible/modules/cloud/packet/packet_device.py [hunk 17]
```diff
@@ -114,7 +112,7 @@
 '''
 
 EXAMPLES = '''
-# All the examples assume that you have your Packet api token in env var PACKET_API_TOKEN.
+# All the examples assume that you have your Packet api token in env var c(PACKET_API_TOKEN).
```
Comment (gundalow) [inline lib/ansible/modules/cloud/packet/packet_device.py]: No need for `C(...)` here as the whole `EXAMPLES` block is already formatted correctly.

File: lib/ansible/modules/cloud/packet/packet_sshkey.py [hunk 18]
```diff
@@ -33,7 +33,7 @@
     choices: ['present', 'absent']
   auth_token:
     description:
-     - Packet api token. You can also supply it in env var PACKET_API_TOKEN.
+     - Packet api token. You can also supply it in env var c(PACKET_API_TOKEN).
```
Comment (gundalow) [inline lib/ansible/modules/cloud/packet/packet_sshkey.py]: Capital `C`.

File: lib/ansible/modules/cloud/packet/packet_sshkey.py [hunk 19]
```diff
@@ -57,7 +57,7 @@
 '''
 
 EXAMPLES = '''
-# All the examples assume that you have your Packet API token in env var PACKET_API_TOKEN.
+# All the examples assume that you have your Packet API token in env var c(PACKET_API_TOKEN).
```
Comment (gundalow) [inline lib/ansible/modules/cloud/packet/packet_sshkey.py]: Remove the `c(...)` here

**Review by gundalow — CHANGES_REQUESTED** [review]
needs_revision

**Comment by t0mk** [comment]
ready_for_review

**Review by gundalow — APPROVED** [review]

**Comment by gundalow** [comment]
Looks like everyone is happy, so merging.

Thanks for the reviews

**Pull request merged**